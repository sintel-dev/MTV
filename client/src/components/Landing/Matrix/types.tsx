export type TagStats = {
  [index: string]: number;
};

export type Scale = {
  maxTagNum: number;
  maxEventNum: number;
  maxScore: number;
};
