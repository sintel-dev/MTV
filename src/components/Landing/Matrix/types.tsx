import { ExperimentDataType } from 'src/model/types';

export type TagStats = {
  [index: string]: number;
};

export type MatrixScale = {
  maxTagNum: number;
  maxEventNum: number;
  maxScore: number;
};

export interface MatrixOption {
  height: number;
  width: number;
  margin: {
    top: number;
    right: number;
    bottom: number;
    left: number;
  };
  matrixHeight: number;
  matrixWidth: number;
}

export interface MatrixProps {
  experiment: ExperimentDataType;
  tagStats: TagStats;
  scale: MatrixScale;
}
