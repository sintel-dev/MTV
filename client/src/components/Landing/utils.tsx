export const colorSchemes = {
  tag: [
    '#FFCD00', // investigate
    '#7CA8FF', // do not investigate
    '#A042FF', // postpone
    '#F64242', // problem
    '#F5FF00', // previously seen
    '#45F642', // normal
    '#C7C7C7', // untagged
  ],
};

// @TODO - later investigation for refactoring (fromIDToTag/fromTagToId)
export const tagSeq = [
  'Investigate',
  'Do not Investigate',
  'Postpone',
  'Problem',
  'Previously seen',
  'Normal',
  'Untagged',
];
const tagClassNames = ['investigate', 'not_investigate', 'postpone', 'problem', 'seen', 'normal', 'untagged'];

export const fromTagToClassName = (tag: string): string => {
  switch (tag) {
    case tagSeq[0]:
      return tagClassNames[0];
    case tagSeq[1]:
      return tagClassNames[1];
    case tagSeq[2]:
      return tagClassNames[2];
    case tagSeq[3]:
      return tagClassNames[3];
    case tagSeq[4]:
      return tagClassNames[4];
    case tagSeq[5]:
      return tagClassNames[5];
    default:
      return 'untagged';
  }
};

export const fromIDtoTag = (id: string): string => {
  switch (id) {
    case '1':
      return tagSeq[0];
    case '2':
      return tagSeq[1];
    case '3':
      return tagSeq[2];
    case '4':
      return tagSeq[3];
    case '5':
      return tagSeq[4];
    case '6':
      return tagSeq[5];
    default:
      return 'untagged';
  }
};

export const fromTagToID = (tag: string): string => {
  switch (tag) {
    case tagSeq[0]:
      return '1';
    case tagSeq[1]:
      return '2';
    case tagSeq[2]:
      return '3';
    case tagSeq[3]:
      return '4';
    case tagSeq[4]:
      return '5';
    case tagSeq[5]:
      return '6';
    default:
      return 'untagged';
  }
};

export const getTagColor = (tag: string): string => {
  let colorIdx: number | undefined;
  for (let i = 0; i < tagSeq.length; i += 1) {
    if (tagSeq[i] === tag) {
      colorIdx = i;
    }
  }
  if (colorIdx === undefined) {
    colorIdx = 6;
  }
  return colorSchemes.tag[colorIdx];
};
