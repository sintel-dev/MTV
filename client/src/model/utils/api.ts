import axios, { AxiosRequestConfig, AxiosInstance } from 'axios';
import Cookies from 'js-cookie';
import { API_URL, SESSION_TOKEN } from './constants';
// import { SESSION_TOKEN } from '@nx-react/session';

import Resource from './restResource';
import {
  DatasetDataType,
  DatasetsResponse,
  DatarunDataType,
  DatarunsResponse,
  ExperimentDataType,
  ExperimentsResponse,
  PipelinesResponse,
  EventsResponse,
  CommentsResponse,
  DataResponse,
} from '../types/index';
import { PipelineDataType } from '../types/pipeline';
import { EventDataType } from '../types/event';
import { CommentDataType } from '../types/comment';

export class RestClient {
  private server: AxiosInstance;

  // default config for http request
  private requestConfig: AxiosRequestConfig = {
    baseURL: API_URL,
    // timeout: 2000,
  };

  public test: Resource<any, any>;

  public datasets: Resource<DatasetDataType, DatasetsResponse>;

  public dataruns: Resource<DatarunDataType, DatarunsResponse>;

  public experiments: Resource<ExperimentDataType, ExperimentsResponse>;

  public pipelines: Resource<PipelineDataType, PipelinesResponse>;

  public events: Resource<EventDataType, EventsResponse>;

  public comments: Resource<CommentDataType, CommentsResponse>;

  public data: Resource<any, DataResponse>;

  /**
   *
   * @param config AxiosRequestConfig
   */
  constructor(config: AxiosRequestConfig) {
    this.requestConfig = { ...this.requestConfig, ...config };
    this.server = axios.create(this.requestConfig);

    // add resources
    this.test = new Resource(this.server, 'test/');
    this.datasets = new Resource(this.server, 'datasets/');
    this.dataruns = new Resource(this.server, 'dataruns/');
    this.experiments = new Resource(this.server, 'experiments/');
    this.pipelines = new Resource(this.server, 'pipelines/');
    this.events = new Resource(this.server, 'events/');
    this.comments = new Resource(this.server, 'comments/');
    this.data = new Resource(this.server, 'data/');
    this.server.interceptors.request.use(this.requestInterceptor);
    this.server.interceptors.response.use(this.responseSuccessInterceptor, this.responseFailInterceptor);
  }

  public requestInterceptor(config) {
    const authHeader = Cookies.get(SESSION_TOKEN);
    if (!authHeader) {
      return config;
    }

    return {
      ...config,
      headers: {
        ...config.headers,
        Authorization: authHeader,
      },
    };
  }

  public responseSuccessInterceptor(response) {
    return response;
  }

  public responseFailInterceptor(error) {
    if (error.response.status === 401) {
      window.location.href = '/logout';
    }

    return Promise.reject(error);
  }
}

export default new RestClient({
  baseURL: 'http://127.0.0.1:3000/api/v1/',
});
