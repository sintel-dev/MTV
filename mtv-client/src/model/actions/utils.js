const baseUrl = '/api/v1/';

export const api = {
    get(url, dataType = 'json') {
        const apiUrl = `${baseUrl}${url}`;
        const promise = fetch(apiUrl, {
            mode: 'cors',
            headers: {
                'Access-Control-Allow-Origin': '*',
            },
        });
        return promise.then(response => {
            if (response.status !== 204) {
                return dataType !== 'json' ? response.text : response.json();
            }
            return null;
        })
        .catch(err => console.error(err));
    },

    // post(url, data) {
    //     const apiUrl = `${baseUrl}${url}`;
    //     return fetch(apiUrl, {
    //         mode: 'cors',
    //         method: 'POST',
    //         headers: {
    //             'Access-Control-Allow-Origin': '*',
    //             'Content-Type': 'application/json',
    //             // 'Authorization': lscache.get('authorization')
    //         },
    //         body: JSON.stringify(data)
    //     });
    // }
};
