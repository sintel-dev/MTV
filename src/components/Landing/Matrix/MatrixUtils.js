export const getMatrixSize = () => {
  const wrapper = document.querySelector('.matrix');
  const width = wrapper.clientWidth - 10;
  const height = wrapper.clientHeight - 10;
  return { width, height };
};
