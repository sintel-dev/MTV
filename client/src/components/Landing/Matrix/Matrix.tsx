import React, { Component } from 'react';
import { connect } from 'react-redux';
import * as d3 from 'd3';
import * as _ from 'lodash';
import { RootState, ExperimentDataType } from '../../../model/types';
import { fromIDtoTag, getTagColor } from '../utils';
import { TagStats, Scale } from './types';
import './Matrix.scss';
import { getMatrixSize } from './MatrixUtils';

// @TODO - gather `tagStats` from selectors/experiment
// import { getProcessedMatrixData } from '../../../model/selectors/experiment'; - later use

type ownProps = {
  experiment: ExperimentDataType;
  tagStats: TagStats;
  scale: Scale;
};
type StateProps = ReturnType<typeof mapState>;
type DispatchProps = ReturnType<typeof mapDispatch>;
type Props = ownProps & StateProps & DispatchProps;
interface LocalState {
  width: number;
  height: number;
}

export class Matrix extends Component<Props, LocalState> {
  constructor(props) {
    super(props);
    this.state = {
      width: 0,
      height: 0,
    };
  }

  componentDidMount() {
    const { width, height } = getMatrixSize();

    this.setState({
      width: Math.max(width, this.option.width),
      height: Math.max(height, this.option.height),
    });
  }

  private option = {
    height: 160, // min height
    width: 180, // min width
    margin: {
      top: 10,
      right: 5,
      bottom: 5,
      left: 2,
    },
    matrixHeight: 100,
    matrixWidth: 120,
  };

  private drawScore(scale, dataruns) {
    const self = this;
    const currentMaxScore = Math.min(6, scale.maxScore);
    let width = self.option.matrixWidth;
    let height = self.option.height - self.option.matrixHeight;
    let bins = new Array(10).fill(0);
    let binNum = bins.length;
    let interval = -1;
    interval = currentMaxScore / binNum;
    let maxCount = 0;
    let outerMaxNum = 0;

    let fx = d3.scaleLinear().range([0, width]);
    let fy = d3.scaleLinear().range([0, height - 16]);

    let area = d3
      .area()
      .x((d) => d[0])
      .y0(0)
      .y1((d) => d[1])
      .curve(d3.curveCardinal);

    dataruns.forEach((currentDatarun) => {
      const { events } = currentDatarun;

      events.forEach((currentEvent) => {
        if (currentEvent.score > currentMaxScore) {
          outerMaxNum += 1;
        } else {
          let idx = Math.trunc(currentEvent.score / interval);
          if (idx === currentEvent.score / interval) {
            idx -= 1;
          }
          bins[idx] += 1;
          maxCount = maxCount < bins[idx] ? bins[idx] : maxCount;
        }
      });
    });

    fx.domain([0, binNum]);
    fy.domain([0, maxCount]);

    let points = _.map(_.range(binNum), (i) => [fx(i + 0.5), -fy(bins[i])]) as [number, number][];
    points.unshift([0, 0]);
    points.push([fx(binNum), -fy(outerMaxNum)]);

    return (
      <g>
        <g transform={`translate(${self.option.width - width}, ${height - 16})`}>
          <path className="area" d={area(points)} />
          <text transform="translate(-10, 12)" className="axis-text" textAnchor="text-anchor">
            0
          </text>
          <text transform={`translate(${width - 20}, 12)`} className="axis-text" textAnchor="text-anchor">
            {currentMaxScore}
          </text>
        </g>
      </g>
    );
  }

  private drawAreaChart(scale) {
    const self = this;
    let width = self.option.matrixWidth;
    let symbol = d3.symbol().size(36);
    let cellWidth = self.option.width - self.option.matrixWidth;
    let cellHeight = self.option.matrixHeight;

    let tScore = false;
    const { experiment } = this.props;
    const { dataruns } = experiment;

    dataruns.forEach((currentDatarun) => {
      const { events } = currentDatarun;

      events.forEach((currentEvent) => {
        if (currentEvent.score > 0) {
          tScore = true;
        }
      });
    });

    return (
      <g>
        <g transform={`translate(${cellWidth}, ${self.option.height - cellHeight})`}>
          <line className="sep-line" x1="0" y1="-14" x2={width} y2="-14" />
          <g transform="translate(4, -8)">
            <path d={symbol.type(d3.symbolTriangle)()} fill="#616365" />
          </g>
          <g transform={`translate(${width - 6}, -8)`}>
            <path d={symbol.type(d3.symbolTriangle)()} fill="#616365" />
          </g>
        </g>
        {tScore && this.drawScore(scale, dataruns)}
      </g>
    );
  }

  private drawCell(cellData, index, scale) {
    const self = this;
    const width = self.option.matrixWidth;
    const rowNum = 6;
    const color = d3.scaleLinear<any>().range(['#A3A4A5', '#43434E']).domain([0, scale.maxEventNum]);

    const fx = d3
      .scaleBand()
      .range([0, width])
      .domain(_.map(d3.range(rowNum), (d) => String(d)))
      .padding(0.1);

    const size = fx.bandwidth();

    return (
      <rect
        className="cell"
        width={size}
        height={size}
        fill={color(cellData[1])}
        x={fx(String(index % rowNum))}
        y={(size + fx.paddingInner() * 1.2) * Math.trunc(index / rowNum)}
        key={Math.random()}
      >
        <title>{`Signal: ${cellData[0]}\nEvents: ${cellData[1]}`}</title>
      </rect>
    );
  }

  private drawMatrix(tagStats, scale) {
    const self = this;
    scale.maxScore = Math.ceil(scale.maxScore);
    let sortedTagStats = _.sortBy(_.toPairs(tagStats), (tag) => +tag[0]);
    let cellWidth = self.option.width - self.option.matrixWidth;
    let cellHeight = self.option.matrixHeight;

    const rh = 7;
    const rm = 3;

    sortedTagStats.push(sortedTagStats[0]);
    sortedTagStats.shift();

    const fx = d3
      .scaleBand()
      .domain(_.map(sortedTagStats, (d) => d[0]))
      .rangeRound([0, cellHeight])
      .paddingInner(0.1);

    const fy = d3
      .scaleLinear()
      .range([0, cellWidth - rh - rm])
      .domain([0, scale.maxTagNum]);

    const curve = d3
      .line()
      .x((d) => d[0])
      .y((d) => d[1])
      .curve(d3.curveBasis);

    const drawBarPath = (tag) => {
      if (tag[1] === 0) {
        return null;
      }

      const [x0, y0, x1, y1] = [-fy(tag[1]) + 1 - rm, fx(tag[0]), -fy(tag[1]) + 1 - rm, fx(tag[0]) + fx.bandwidth()];
      const [xm0, ym0, xm1, ym1] = [x0 - rh, y0, x1 - rh, y1];

      const path = curve([
        [x0, y0],
        [xm0, ym0],
        [xm1, ym1],
        [x1, y1],
      ]);
      return path;
    };

    const cells: [string, number][] = _.chain(self.props.experiment.dataruns)

      .map((d) => [d.signal, d.events.length])
      .sortBy((d) => d[0])
      .value() as [string, number][];

    return (
      <g transform={`translate(${self.option.margin.left}, ${self.option.margin.top})`}>
        <g transform={`translate(${cellWidth}, ${self.option.height - cellHeight})`}>
          {sortedTagStats.map((currentTag: Array<any>) => (
            <g key={Math.random()}>
              <rect
                className="bar"
                x={-fy(currentTag[1]) - rm}
                y={fx(currentTag[0])}
                width={fy(currentTag[1])}
                height={fx.bandwidth()}
                fill={getTagColor(fromIDtoTag(currentTag[0]))}
              >
                <title>
                  {fromIDtoTag(currentTag[0])}: {currentTag[1]}
                </title>
              </rect>
              {drawBarPath(currentTag) && (
                <path className="bar-hat" d={drawBarPath(currentTag)} fill={getTagColor(fromIDtoTag(currentTag[0]))} />
              )}
            </g>
          ))}
          <line className="sep-line" x1={0} y1={0} x2={0} y2={cellHeight} />
        </g>
        <g transform={`translate(${cellWidth}, ${self.option.height - cellHeight})`}>
          {cells && cells.map((cell, index) => this.drawCell(cell, index, scale))}
        </g>
        {this.drawAreaChart(scale)}
      </g>
    );
  }

  render() {
    const { width, height } = this.state;
    const { tagStats, scale } = this.props;

    return (
      <div data-eid={this.props.experiment.id} className="matrix">
        <svg width={width} height={height} className="matrix-svg">
          {this.drawMatrix(tagStats, scale)}
        </svg>
      </div>
    );
  }
}

const mapState = () => ({});

const mapDispatch = () => ({});

export default connect<StateProps, DispatchProps, ownProps, RootState>(mapState, mapDispatch)(Matrix);
