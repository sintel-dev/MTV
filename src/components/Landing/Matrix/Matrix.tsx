import React, { Component } from 'react';
import * as d3 from 'd3';
import * as _ from 'lodash';
import { DatarunDataType } from 'src/model/types/datarun';
import { fromIDtoTag, getTagColor } from '../utils';
import { MatrixOption, MatrixProps } from './types';
import { getMatrixSize } from './MatrixUtils';
import './Matrix.scss';

interface LocalState {
  width: number;
  height: number;
}

export class Matrix extends Component<MatrixProps, LocalState> {
  constructor(props) {
    super(props);
    this.state = {
      width: 0,
      height: 0,
    };
  }

  componentDidMount() {
    const { width, height } = getMatrixSize();

    this.setState({
      width: Math.max(width, this.option.width),
      height: Math.max(height, this.option.height),
    });
  }

  private option: MatrixOption = {
    height: 160, // min height
    width: 180, // min width
    margin: {
      top: 10,
      right: 5,
      bottom: 5,
      left: 2,
    },
    matrixHeight: 100,
    matrixWidth: 120,
  };

  private drawScore(scale, dataruns: DatarunDataType[]) {
    const currentMaxScore: number = Math.min(6, scale.maxScore);
    let width: number = this.option.matrixWidth;
    let height: number = this.option.height - this.option.matrixHeight;
    let bins: Array<number> = new Array(10).fill(0);
    let binNum: number = bins.length;
    let interval = -1;
    interval = currentMaxScore / binNum;
    let maxCount = 0;
    let outerMaxNum = 0;

    let fx = d3.scaleLinear().range([0, width]);
    let fy = d3.scaleLinear().range([0, height - 16]);

    let area = d3
      .area()
      .x((d) => d[0])
      .y0(0)
      .y1((d) => d[1])
      .curve(d3.curveCardinal);

    dataruns.forEach((currentDatarun) => {
      const { events } = currentDatarun;
      events.forEach((currentEvent) => {
        if (currentEvent.score > currentMaxScore) {
          outerMaxNum += 1;
        } else {
          let idx: number = Math.trunc(currentEvent.score / interval);
          if (idx === currentEvent.score / interval) {
            idx -= 1;
          }
          bins[idx] += 1;
          maxCount = maxCount < bins[idx] ? bins[idx] : maxCount;
        }
      });
    });

    fx.domain([0, binNum]);
    fy.domain([0, maxCount]);

    let points: [number, number][] = _.map(_.range(binNum), (i) => [fx(i + 0.5), -fy(bins[i])]);
    points.unshift([0, 0]);
    points.push([fx(binNum), -fy(outerMaxNum)]);

    return (
      <g>
        <g transform={`translate(${this.option.width - width}, ${height - 16})`}>
          <path className="area" d={area(points)} />
          <text transform="translate(-10, 12)" className="axis-text" textAnchor="text-anchor">
            0
          </text>
          <text transform={`translate(${width - 20}, 12)`} className="axis-text" textAnchor="text-anchor">
            {currentMaxScore}
          </text>
        </g>
      </g>
    );
  }

  private drawAreaChart(scale) {
    const { height, matrixWidth, matrixHeight } = this.option;
    let width: number = matrixWidth;
    let symbol = d3.symbol().size(36);
    let cellWidth: number = this.option.width - matrixWidth;
    let cellHeight: number = matrixHeight;

    let tScore = false;
    const { experiment } = this.props;
    const { dataruns } = experiment;

    dataruns.forEach((currentDatarun) => {
      const { events } = currentDatarun;

      events.forEach((currentEvent) => {
        if (currentEvent.score > 0) {
          tScore = true;
        }
      });
    });

    return (
      <g>
        <g transform={`translate(${cellWidth}, ${height - cellHeight})`}>
          <line className="sep-line" x1="0" y1="-14" x2={width} y2="-14" />
          <g transform="translate(4, -8)">
            <path d={symbol.type(d3.symbolTriangle)()} fill="#616365" />
          </g>
          <g transform={`translate(${width - 6}, -8)`}>
            <path d={symbol.type(d3.symbolTriangle)()} fill="#616365" />
          </g>
        </g>
        {tScore && this.drawScore(scale, dataruns)}
      </g>
    );
  }

  private drawCell(cellData, index, scale) {
    const self = this;
    const width = self.option.matrixWidth;
    const rowNum = 6;
    const color = d3.scaleLinear<any>().range(['#A3A4A5', '#43434E']).domain([0, scale.maxEventNum]);

    const fx = d3
      .scaleBand()
      .range([0, width])
      .domain(_.map(d3.range(rowNum), (d) => String(d)))
      .padding(0.1);

    const size = fx.bandwidth();

    return (
      <rect
        className="cell"
        width={size}
        height={size}
        fill={color(cellData[1])}
        x={fx(String(index % rowNum))}
        y={(size + fx.paddingInner() * 1.2) * Math.trunc(index / rowNum)}
        key={Math.random()}
      >
        <title>{`Signal: ${cellData[0]}\nEvents: ${cellData[1]}`}</title>
      </rect>
    );
  }

  private drawMatrix(tagStats, scale) {
    const { dataruns } = this.props.experiment;
    const { margin, height } = this.option;
    scale.maxScore = Math.ceil(scale.maxScore);
    let sortedTagStats = _.sortBy(_.toPairs(tagStats), (tag) => +tag[0]);
    let cellWidth = this.option.width - this.option.matrixWidth;
    let cellHeight = this.option.matrixHeight;

    const rh = 7;
    const rm = 3;

    sortedTagStats.push(sortedTagStats[0]);
    sortedTagStats.shift();

    const fx = d3
      .scaleBand()
      .domain(_.map(sortedTagStats, (d) => d[0]))
      .rangeRound([0, cellHeight])
      .paddingInner(0.1);

    const fy = d3
      .scaleLinear()
      .range([0, cellWidth - rh - rm])
      .domain([0, scale.maxTagNum]);

    const curve = d3
      .line()
      .x((d) => d[0])
      .y((d) => d[1])
      .curve(d3.curveBasis);

    const drawBarPath = (tag) => {
      if (tag[1] === 0) {
        return null;
      }

      const [x0, y0, x1, y1] = [
        -fy(tag[1]) + 1 - rm,
        fx(tag[0]),
        -fy(tag[1]) + 1 - rm,
        fx(tag[0]) + fx.bandwidth(),
      ] as [number, number, number, number];

      const [xm0, ym0, xm1, ym1] = [x0 - rh, y0, x1 - rh, y1] as [number, number, number, number];

      const path = curve([
        [x0, y0],
        [xm0, ym0],
        [xm1, ym1],
        [x1, y1],
      ]);
      return path;
    };

    const cells: [string, number][] = _.chain(dataruns)
      .map((currentDatarun) => [currentDatarun.signal, currentDatarun.events.length])
      .sortBy((currentDatarun) => currentDatarun[0])
      .value() as [string, number][];

    return (
      <g transform={`translate(${margin.left}, ${margin.top})`}>
        <g transform={`translate(${cellWidth}, ${height - cellHeight})`}>
          {sortedTagStats.map((currentTag: Array<any>) => (
            <g key={Math.random()}>
              <rect
                className="bar"
                x={-fy(currentTag[1]) - rm}
                y={fx(currentTag[0])}
                width={fy(currentTag[1])}
                height={fx.bandwidth()}
                fill={getTagColor(fromIDtoTag(currentTag[0]))}
              >
                <title>
                  {fromIDtoTag(currentTag[0])}: {currentTag[1]}
                </title>
              </rect>
              {drawBarPath(currentTag) && (
                <path className="bar-hat" d={drawBarPath(currentTag)} fill={getTagColor(fromIDtoTag(currentTag[0]))} />
              )}
            </g>
          ))}
          <line className="sep-line" x1={0} y1={0} x2={0} y2={cellHeight} />
        </g>
        <g transform={`translate(${cellWidth}, ${height - cellHeight})`}>
          {cells && cells.map((cell, index: number) => this.drawCell(cell, index, scale))}
        </g>
        {this.drawAreaChart(scale)}
      </g>
    );
  }

  render() {
    const { width, height } = this.state;
    const { tagStats, scale, experiment } = this.props;

    return (
      <div data-eid={experiment.id} className="matrix">
        <svg width={width} height={height} className="matrix-svg">
          {this.drawMatrix(tagStats, scale)}
        </svg>
      </div>
    );
  }
}

export default Matrix;
