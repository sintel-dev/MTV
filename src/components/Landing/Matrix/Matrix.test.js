import React from 'react';
import { Matrix } from './Matrix';
import { renderWithStore, TestWrapper } from '../../../tests/utils';

describe('Testing the matrix component', () => {
  it('Should render matrix component without crashing', () => {
    const matrixProps = {
      experiment: {
        id: '5da7cc6376e3e19307d0db64',
        project: 'SMAP',
        dataset: 'SMAP_set1',
        date_creation: '2019-10-17T02:05:23.503000',
        created_by: null,
        pipeline: 'lstm',
        dataruns: [
          {
            id: '5da7cc6576e3e19307d0db65',
            signal: 'T-1',
            events: [{ start_time: 1228219200, stop_time: 1229299200, score: 0.0336717714726245, tag: null }],
          },
        ],
      },
      tagStats: { 0: 30, 1: 0, 2: 0, 3: 0, 4: 0, 5: 0, 6: 0 },
      scale: { maxTagNum: 64, maxEventNum: 10, maxScore: 1.5598205662590257 },
    };
    const mountedMatrix = renderWithStore(
      {},
      <TestWrapper>
        <Matrix {...matrixProps} />
      </TestWrapper>,
    );

    expect(mountedMatrix).toMatchSnapshot();
  });
});
