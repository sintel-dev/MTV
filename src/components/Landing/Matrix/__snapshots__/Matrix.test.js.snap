// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing the matrix component Should render matrix component without crashing 1`] = `
<div
  class="matrix"
  data-eid="5da7cc6376e3e19307d0db64"
>
  <svg
    class="matrix-svg"
    height="0"
    width="0"
  >
    <g
      transform="translate(2, 10)"
    >
      <g
        transform="translate(60, 60)"
      >
        <g>
          <rect
            class="bar"
            fill="#FF9100"
            height="13"
            width="0"
            x="-3"
            y="2"
          >
            <title>
              Investigate: 0
            </title>
          </rect>
        </g>
        <g>
          <rect
            class="bar"
            fill="#4684FF"
            height="13"
            width="0"
            x="-3"
            y="16"
          >
            <title>
              Do not Investigate: 0
            </title>
          </rect>
        </g>
        <g>
          <rect
            class="bar"
            fill="#8446C2"
            height="13"
            width="0"
            x="-3"
            y="30"
          >
            <title>
              Postpone: 0
            </title>
          </rect>
        </g>
        <g>
          <rect
            class="bar"
            fill="#C24D4D"
            height="13"
            width="0"
            x="-3"
            y="44"
          >
            <title>
              Problem: 0
            </title>
          </rect>
        </g>
        <g>
          <rect
            class="bar"
            fill="#F2FF00"
            height="13"
            width="0"
            x="-3"
            y="58"
          >
            <title>
              Previously seen: 0
            </title>
          </rect>
        </g>
        <g>
          <rect
            class="bar"
            fill="#6BC969"
            height="13"
            width="0"
            x="-3"
            y="72"
          >
            <title>
              Normal: 0
            </title>
          </rect>
        </g>
        <g>
          <rect
            class="bar"
            fill="#D5D5D5"
            height="13"
            width="23.4375"
            x="-26.4375"
            y="86"
          >
            <title>
              untagged: 30
            </title>
          </rect>
          <path
            class="bar-hat"
            d="M-25.4375,86L-26.604166666666668,86C-27.770833333333332,86,-30.104166666666668,86,-31.270833333333332,88.16666666666667C-32.4375,90.33333333333333,-32.4375,94.66666666666667,-31.270833333333332,96.83333333333333C-30.104166666666668,99,-27.770833333333332,99,-26.604166666666668,99L-25.4375,99"
            fill="#D5D5D5"
          />
        </g>
        <line
          class="sep-line"
          x1="0"
          x2="0"
          y1="0"
          y2="100"
        />
      </g>
      <g
        transform="translate(60, 60)"
      >
        <rect
          class="cell"
          fill="rgb(153, 154, 156)"
          height="17.704918032786882"
          width="17.704918032786882"
          x="1.9672131147541023"
          y="0"
        >
          <title>
            Signal: T-1
Events: 1
          </title>
        </rect>
      </g>
      <g>
        <g
          transform="translate(60, 60)"
        >
          <line
            class="sep-line"
            x1="0"
            x2="120"
            y1="-14"
            y2="-14"
          />
          <g
            transform="translate(4, -8)"
          >
            <path
              d="M0,-5.2642960518099695L4.559014113909555,2.6321480259049848L-4.559014113909555,2.6321480259049848Z"
              fill="#616365"
            />
          </g>
          <g
            transform="translate(114, -8)"
          >
            <path
              d="M0,-5.2642960518099695L4.559014113909555,2.6321480259049848L-4.559014113909555,2.6321480259049848Z"
              fill="#616365"
            />
          </g>
        </g>
        <g>
          <g
            transform="translate(60, 44)"
          >
            <path
              class="area"
              d="M0,0C0,0,3,-44,6,-44C9,-44,14,-7.333333333333333,18,0C22,7.333333333333333,26,0,30,0C34,0,38,0,42,0C46,0,50,0,54,0C58,0,62,0,66,0C70,0,74,0,78,0C82,0,86,0,90,0C94,0,98,0,102,0C106,0,111,0,114,0C117,0,120,0,120,0L120,0C120,0,117,0,114,0C111,0,106,0,102,0C98,0,94,0,90,0C86,0,82,0,78,0C74,0,70,0,66,0C62,0,58,0,54,0C50,0,46,0,42,0C38,0,34,0,30,0C26,0,22,0,18,0C14,0,9,0,6,0C3,0,0,0,0,0Z"
            />
            <text
              class="axis-text"
              text-anchor="text-anchor"
              transform="translate(-10, 12)"
            >
              0
            </text>
            <text
              class="axis-text"
              text-anchor="text-anchor"
              transform="translate(100, 12)"
            >
              2
            </text>
          </g>
        </g>
      </g>
    </g>
  </svg>
</div>
`;
